File: src/main/java/frc/robot/io/TalonFXMotorIOTunable.java
package frc.robot.io;

import com.ctre.phoenix6.configs.CANcoderConfiguration;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.controls.Follower;
import frc.robot.util.TalonFXTunableHelper;
import java.util.Optional;

/** Talon FX IO wrapper that publishes runtime tuning controls for CTRE configs. */
public class TalonFXMotorIOTunable extends TalonFXMotorIO {
  private final TalonFXTunableHelper tunableHelper;

  /** Creates a tunable Talon FX with no follower or CANcoder. */
  public TalonFXMotorIOTunable(int canID, TalonFXConfiguration config, String motorName) {
    this(canID, config, motorName, Optional.empty(), Optional.empty());
  }

  /** Creates a tunable Talon FX configured as a follower. */
  public TalonFXMotorIOTunable(
      int canID, TalonFXConfiguration config, String motorName, Follower followerConfig) {
    this(canID, config, motorName, Optional.empty(), Optional.ofNullable(followerConfig));
  }

  /** Creates a tunable Talon FX using external CANcoder feedback. */
  public TalonFXMotorIOTunable(
      int canID,
      TalonFXConfiguration config,
      String motorName,
      CANcoderConfiguration canCoderConfig) {
    this(canID, config, motorName, Optional.ofNullable(canCoderConfig), Optional.empty());
  }

  /** Full constructor for selecting optional CANcoder/follower config. */
  public TalonFXMotorIOTunable(
      int canID,
      TalonFXConfiguration config,
      String motorName,
      Optional<CANcoderConfiguration> canCoderConfig,
      Optional<Follower> followerConfig) {
    super(canID, config, motorName, canCoderConfig, followerConfig);

    String subsystem = extractSubsystem(motorName);
    String baseKey = subsystem + "/Tuning/TalonFX/" + motorName + "/";
    tunableHelper = new TalonFXTunableHelper(motor, baseKey, config, motorName);
  }

  /**
   * Extracts the subsystem name from the first word of the motor name.
   *
   * <p>Examples: "Intake Roller" -> "Intake", "Shooter Hood" -> "Shooter".
   */
  private static String extractSubsystem(String motorName) {
    int space = motorName.indexOf(' ');
    return (space >= 0) ? motorName.substring(0, space) : motorName;
  }
}

File: src/main/cpp/io/TalonFXMotorIOTunable.cpp
#include "io/TalonFXMotorIOTunable.h"
#include "util/TalonFXTunableHelper.h"

using namespace ctre::phoenix6;

TalonFXMotorIOTunable::TalonFXMotorIOTunable(
    int canID, configs::TalonFXConfiguration config, std::string motorName)
    : TalonFXMotorIOTunable(canID, std::move(config), std::move(motorName),
                            std::nullopt, std::nullopt) {}

TalonFXMotorIOTunable::TalonFXMotorIOTunable(
    int canID, configs::TalonFXConfiguration config, std::string motorName,
    controls::Follower followerConfig)
    : TalonFXMotorIOTunable(canID, std::move(config), std::move(motorName),
                            std::nullopt, std::make_optional(std::move(followerConfig))) {}

TalonFXMotorIOTunable::TalonFXMotorIOTunable(
    int canID, configs::TalonFXConfiguration config, std::string motorName,
    configs::CANcoderConfiguration canCoderConfig)
    : TalonFXMotorIOTunable(canID, std::move(config), std::move(motorName),
                            std::make_optional(std::move(canCoderConfig)), std::nullopt) {}

TalonFXMotorIOTunable::TalonFXMotorIOTunable(
    int canID, configs::TalonFXConfiguration config, std::string motorName,
    std::optional<configs::CANcoderConfiguration> canCoderConfig,
    std::optional<controls::Follower> followerConfig)
    : TalonFXMotorIO(canID, std::move(config), std::move(motorName),
                     std::move(canCoderConfig), std::move(followerConfig)) {
  std::string subsystem = ExtractSubsystem(m_motorName);
  std::string baseKey = subsystem + "/Tuning/TalonFX/" + m_motorName + "/";
  m_tunableHelper = std::make_unique<TalonFXTunableHelper>(
      m_motor, baseKey, m_config, m_motorName);
}

std::string TalonFXMotorIOTunable::ExtractSubsystem(const std::string& motorName) {
  auto space = motorName.find(' ');
  return (space != std::string::npos) ? motorName.substr(0, space) : motorName;
}
